[BUG] Null parameter overrides default value

### Existing rule

Azure.PublicIP.StandardSKU

### Description of the issue

I'm trying to fix the WAF-Reliability (Azure.Pillar.Reliability) of the AVM module `avm/res/network/bastion-host` and the tests are failing on following rules:
> AZR-000158: Azure.PublicIP.StandardSKU. Public IP addresses should be deployed with Standard SKU for production workloads.
> AZR-000395: Azure.PublicIP.MigrateStandard. Use the Standard SKU for Public IP addresses as the Basic SKU will be retired.

Scenario:
- The Bastion module is cross-referencing the module `br/public:avm/res/network/public-ip-address:0.3.2` for Public IP creation, which has a default `skuName` set to `'Standard'`. In the failing test I'm not setting the Public IP Sku explicitly, I'm relying on the default value of the cross-referenced Public IP module. The test is failing: https://github.com/krbar/bicep-registry-modules/actions/runs/8520500847/job/23336745635
- Even if I explicitly add the `skuName: 'Standard'`, which should not be necessary as this is the default, the test failed anyway: https://github.com/krbar/bicep-registry-modules/actions/runs/8382290824
- When I skip the PSRule validation and perform the deployment, the SKU of the deployed Public IP in is 'Standard'.
- the Basic SKU is not used, but the rule `Azure.PublicIP.MigrateStandard` fails as well

  CC: @eriqua  @jtracey93

### Error messages

```
[info] Using Version: 2.9.0
[info] Using Action: __self_2
[info] Using Workspace: /home/runner/work/bicep-registry-modules/bicep-registry-modules
[info] Using PWD: /home/runner/work/bicep-registry-modules/bicep-registry-modules
[info] Using Path: /home/runner/work/bicep-registry-modules/bicep-registry-modules
[info] Using Source: /home/runner/work/bicep-registry-modules/bicep-registry-modules/avm/utilities/pipelines/staticValidation/psrule/.ps-rule/
[info] Using Baseline: Azure.Pillar.Reliability
[info] Using Conventions: 
[info] Using InputType: repository
[info] Using InputPath: /home/runner/work/bicep-registry-modules/bicep-registry-modules/avm/res/network/bastion-host/tests/e2e/defaults/main.test.bicep
[info] Using Option: /home/runner/work/bicep-registry-modules/bicep-registry-modules/avm/utilities/pipelines/staticValidation/psrule/ps-rule.yaml
[info] Using Outcome: 
[info] Using OutputFormat: Csv
[info] Using OutputPath: avm/res/network/bastion-host/tests/e2e/defaults/main.test.bicep-PSRule-output.csv
[info] Using Summary: false

---
    ____  _____ ____        __
   / __ \/ ___// __ \__  __/ /__
  / /_/ /\__ \/ /_/ / / / / / _ \
 / ____/___/ / _, _/ /_/ / /  __/
/_/    /____/_/ |_|\__,_/_/\___/

Using PSRule v2.9.0
Using PSRule.Rules.Azure v1.35.0

----------------------------
Explore documentation: https://aka.ms/ps-rule
Contribute and find source: https://github.com/microsoft/PSRule
Report issues: https://github.com/microsoft/PSRule/issues
PSRule.Rules.Azure: https://aka.ms/ps-rule-azure
----------------------------

From repository: https://github.com/krbar/bicep-registry-modules
  on : users/krbar/bastionWAF
  at : 808f50005afdbcc7f503e3bc92112371db0e13fe

 -> ***nbhmin001-pip : Microsoft.Network/publicIPAddresses [0/1]

    [FAIL] Azure.PublicIP.StandardSKU (AZR-000158)
    | Template: avm/res/network/bastion-host/tests/e2e/defaults/main.test.bicep:1107:40

    | RECOMMEND:
    | Consider using Standard SKU for Public IP addresses deployed in production.

    | REASON:
    | - Path sku.name: The field 'sku.name' does not exist.

    | HELP:
    | - https://azure.github.io/PSRule.Rules.Azure/en/rules/Azure.PublicIP.StandardSKU/

Error: AZR-000158: ***nbhmin001-pip failed Azure.PublicIP.StandardSKU. Public IP addresses should be deployed with Standard SKU for production workloads.


Rules processed: 1, failed: 1, errored: 0
Run krbar/bicep-registry-modules/85[205](https://github.com/krbar/bicep-registry-modules/actions/runs/8520500847/job/23336745635#step:4:210)00847 completed in 00:00:05.9216107
Output written to the following file: '/home/runner/work/bicep-registry-modules/bicep-registry-modules/avm/res/network/bastion-host/tests/e2e/defaults/main.test.bicep-PSRule-output.csv'
Error: One or more assertions failed. One or more rules reported failure.

---
Error: Process completed with exit code 1.
```


### Reproduction

Perform the PS Rule validation on the file [avm/res/network/bastion-host/tests/e2e/defaults/main.test.bicep](https://github.com/krbar/bicep-registry-modules/blob/users/krbar/bastionWAF/avm/res/network/bastion-host/tests/e2e/defaults/main.test.bicep)

Link to the module root folder: https://github.com/krbar/bicep-registry-modules/tree/users/krbar/bastionWAF/avm/res/network/bastion-host


### Version of PSRule

2.9.0

### Version of PSRule for Azure

1.35.0

### Additional context

_No response_

--- a/docs/CHANGELOG-v1.md
+++ b/docs/CHANGELOG-v1.md
@@ -32,6 +32,12 @@ See [upgrade notes][1] for helpful information when upgrading from previous vers
 
 ## Unreleased
 
+What's changed since v1.35.0:
+
+- Bug fixes:
+  - Fixed null parameter overrides default value by @BernieWhite.
+    [#2795](https://github.com/Azure/PSRule.Rules.Azure/issues/2795)
+
 ## v1.35.0
 
 What's changed since v1.34.2:

--- a/src/PSRule.Rules.Azure/Data/Template/TemplateVisitor.cs
+++ b/src/PSRule.Rules.Azure/Data/Template/TemplateVisitor.cs
@@ -352,6 +352,9 @@ namespace PSRule.Rules.Azure.Data.Template
 
             private void AddParameterAssignment(string name, JToken value)
             {
+                if (value == null || value.Type == JTokenType.Null || value.Type == JTokenType.Undefined)
+                    return;
+
                 _ParameterAssignments.Add(name, value);
             }
 

--- a/tests/PSRule.Rules.Azure.Tests/TemplateVisitorTests.cs
+++ b/tests/PSRule.Rules.Azure.Tests/TemplateVisitorTests.cs
@@ -1099,9 +1099,21 @@ namespace PSRule.Rules.Azure
         /// Test case for https://github.com/Azure/PSRule.Rules.Azure/issues/2751.
         /// </summary>
         [Fact]
-        public void Index_into_mock_output_object()
+        public void ProcessTemplate_WhenIndexIntoMock_ShouldReturnMock()
         {
             var resources = ProcessTemplate(GetSourcePath("Tests.Bicep.35.json"), null, out _);
+        }
+
+        /// <summary>
+        /// Test case for https://github.com/Azure/PSRule.Rules.Azure/issues/2795.
+        /// </summary>
+        [Fact]
+        public void ProcessTemplate_WhenParameterNullWithDefault_ShouldUseDefault()
+        {
+            var resources = ProcessTemplate(GetSourcePath("Tests.Bicep.27.json"), null, out _);
+
+            var actual = resources.FirstOrDefault(r => r["type"].Value<string>() == "Microsoft.Storage/storageAccounts");
+            Assert.Equal("Standard_LRS", actual["sku"]["name"].Value<string>());
         }
 
         #region Helper methods

--- a/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.27.bicep
+++ b/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.27.bicep
@@ -3,6 +3,9 @@
 
 module child 'Tests.Bicep.27.child.bicep' = {
   name: 'child'
+  params: {
+    skuName: null
+  }
 }
 
 output childFromFor string = child.outputs.fromFor

--- a/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.27.child.bicep
+++ b/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.27.child.bicep
@@ -1,6 +1,7 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT License.
 
+param skuName string = 'Standard_LRS'
 param minTLSVersion string?
 param corsRules corsRule
 
@@ -17,11 +18,13 @@ param accessPolicies array?
 @secure()
 param secrets object?
 
-var formattedAccessPolicies = [for accessPolicy in (accessPolicies ?? []): {
-  objectId: accessPolicy.objectId
-  tenantId: contains(accessPolicy, 'tenantId') ? accessPolicy.tenantId : tenant().tenantId
-  permissions: {}
-}]
+var formattedAccessPolicies = [
+  for accessPolicy in (accessPolicies ?? []): {
+    objectId: accessPolicy.objectId
+    tenantId: contains(accessPolicy, 'tenantId') ? accessPolicy.tenantId : tenant().tenantId
+    permissions: {}
+  }
+]
 
 var secretList = secrets.?secureList ?? []
 
@@ -30,7 +33,7 @@ resource storage 'Microsoft.Storage/storageAccounts@2023-01-01' = {
   #disable-next-line no-loc-expr-outside-params
   location: resourceGroup().location
   sku: {
-    name: 'Standard_LRS'
+    name: skuName
   }
   kind: 'StorageV2'
   properties: {
@@ -61,12 +64,14 @@ resource kv 'Microsoft.KeyVault/vaults@2023-07-01' = {
   }
 }
 
-resource kvSecret 'Microsoft.KeyVault/vaults/secrets@2023-07-01' = [for item in secretList: {
-  name: item.name
-  properties: {
-    value: item.value
+resource kvSecret 'Microsoft.KeyVault/vaults/secrets@2023-07-01' = [
+  for item in secretList: {
+    name: item.name
+    properties: {
+      value: item.value
+    }
   }
-}]
+]
 
 resource storageAccount_objectReplicationPolicy 'Microsoft.Storage/storageAccounts/objectReplicationPolicies@2022-09-01' = {
   name: 'default'
@@ -85,22 +90,24 @@ resource storageAccount_objectReplicationPolicy 'Microsoft.Storage/storageAccoun
   }
 }
 
-resource storageAccount_objectReplicationPolicyItems 'Microsoft.Storage/storageAccounts/objectReplicationPolicies@2022-09-01' = [for (item, index) in [ 1 ]: {
-  name: 'default${index}'
-  parent: storage
-  properties: {
-    sourceAccount: 'sourceId'
-    destinationAccount: 'destId'
-    rules: [
-      {
-        ruleId: null
-        sourceContainer: 'source'
-        destinationContainer: 'dest'
-        filters: null
-      }
-    ]
+resource storageAccount_objectReplicationPolicyItems 'Microsoft.Storage/storageAccounts/objectReplicationPolicies@2022-09-01' = [
+  for (item, index) in [1]: {
+    name: 'default${index}'
+    parent: storage
+    properties: {
+      sourceAccount: 'sourceId'
+      destinationAccount: 'destId'
+      rules: [
+        {
+          ruleId: null
+          sourceContainer: 'source'
+          destinationContainer: 'dest'
+          filters: null
+        }
+      ]
+    }
   }
-}]
+]
 
 output policyId string = storageAccount_objectReplicationPolicy.properties.policyId
 output ruleIds string[] = map(storageAccount_objectReplicationPolicy.properties.rules, rule => rule.ruleId)

--- a/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.27.json
+++ b/tests/PSRule.Rules.Azure.Tests/Tests.Bicep.27.json
@@ -4,8 +4,8 @@
   "metadata": {
     "_generator": {
       "name": "bicep",
-      "version": "0.23.1.45101",
-      "templateHash": "1063354008601109842"
+      "version": "0.26.54.24096",
+      "templateHash": "7737727980270162371"
     }
   },
   "resources": [
@@ -18,6 +18,11 @@
           "scope": "inner"
         },
         "mode": "Incremental",
+        "parameters": {
+          "skuName": {
+            "value": null
+          }
+        },
         "template": {
           "$schema": "https://schema.management.azure.com/schemas/2019-04-01/deploymentTemplate.json#",
           "languageVersion": "2.0",
@@ -25,8 +30,8 @@
           "metadata": {
             "_generator": {
               "name": "bicep",
-              "version": "0.23.1.45101",
-              "templateHash": "11238883118622659404"
+              "version": "0.26.54.24096",
+              "templateHash": "14665743710501821887"
             }
           },
           "definitions": {
@@ -68,6 +73,10 @@
             }
           },
           "parameters": {
+            "skuName": {
+              "type": "string",
+              "defaultValue": "Standard_LRS"
+            },
             "minTLSVersion": {
               "type": "string",
               "nullable": true
@@ -118,7 +127,7 @@
               "name": "test",
               "location": "[resourceGroup().location]",
               "sku": {
-                "name": "Standard_LRS"
+                "name": "[parameters('skuName')]"
               },
               "kind": "StorageV2",
               "properties": {

